src/transaction/flashback.c

```diff
@@ -0,0 +1,416 @@
+}
+
+/*
+ * flashback_fill_dml_summary - fill the dml information into summary entry
+ *
+ * thread_p (in)             : thread entry
+ * summary_entry (in/out)    : flashback summary entry
+ * classoid (in)             : OID of DML target class
+ * rec_type(in)              : record type that can be classified as DML types
+ */
+
+static void
+flashback_fill_dml_summary (FLASHBACK_SUMMARY_ENTRY * summary_entry, OID classoid, SUPPLEMENT_REC_TYPE rec_type)
+{
+  assert (summary_entry != NULL);
+
+  bool found = false;
+  int i = 0;
+
+  /* fill in the summary_entry info entry
+   * 1. number of class and put the classoid into the classlist
+   * 2. number of DML operation */
+
+  for (i = 0; i < summary_entry->num_class; i++)
+    {
+      if (OID_EQ (&summary_entry->classlist[i], &classoid))
+	{
+	  found = true;
+	  break;
+	}
+    }
+
+  if (!found)
+    {
+      summary_entry->num_class += 1;
+      COPY_OID (&summary_entry->classlist[i], &classoid);
+    }
+
+  switch (rec_type)
+    {
+    case LOG_SUPPLEMENT_INSERT:
+    case LOG_SUPPLEMENT_TRIGGER_INSERT:
+      summary_entry->num_insert++;
+      break;
+    case LOG_SUPPLEMENT_UPDATE:
+    case LOG_SUPPLEMENT_TRIGGER_UPDATE:
+      summary_entry->num_update++;
+      break;
+    case LOG_SUPPLEMENT_DELETE:
+    case LOG_SUPPLEMENT_TRIGGER_DELETE:
+      summary_entry->num_delete++;
+      break;
+    default:
+      assert (false);
+      break;
+    }
+
+  return;
+}
+
+/*
+ * flashback_cleanup - dealloc memory of summary entries iterating summary list
+ *
+ * thread_p (in)         : thread entry
+ * context (in/out)      : flashback summary context that contains summary list
+ */
+
+void
+flashback_cleanup (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  /* *INDENT-OFF* */
+  for (auto iter:context->summary_list)
+    {
+      if (iter.second != NULL)
+        {
+	  db_private_free_and_init (thread_p, iter.second);
+        }
+    }
+  /* *INDENT-ON* */
+}
+
+/*
+ * flashback_make_summary - read log volumes within the range of start_lsa and end_lsa then make summary information
+ *
+ * return                : error code
+ * thread_p (in)         : thread entry
+ * context (in/out)      : flashback summary context that contains range of travering logs, filter (user, class), and summary list
+ */
+
+int
+flashback_make_summary_list (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  LOG_LSA cur_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA next_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA process_lsa = LSA_INITIALIZER;
+
+  LOG_RECORD_HEADER *log_rec_header;
+
+  int trid;
+
+  time_t current_time = 0;
+
+  FLASHBACK_SUMMARY_ENTRY *summary_entry = NULL;
+
+  int error = NO_ERROR;
+
+  LOG_RECTYPE log_type;
+
+  char *supplement_data = NULL;
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  LSA_COPY (&process_lsa, &(context->end_lsa));
```

Comment:
I think it would be nice if there is an exception handling code after this line like 'if (LSA_ISNULL(&process_lsa)) goto error'.

Reply:
I think context->end_lsa can not be null, but I will add `assert (!LSA_ISNULL (&context->end_lsa))`

Reply:
OK.